lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp

```diff
@@ -0,0 +1,181 @@
+    if (!fas) {
+      continue;
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Partial Apply: " << *pai);
+    LLVM_DEBUG(llvm::dbgs() << "Full Apply Site: " << *fas.getInstruction());
+    auto *fasi = dyn_cast<ApplyInst>(fas.getInstruction());
+    if (!fasi) {
+      continue;
+    }
+
+    auto fasArgs = fas.getArguments();
+    if (!areAllValuesTrivial(fasArgs, *function)) {
+      continue;
+    }
+
+    SmallVector<SILValue, 8> argsVec;
+    llvm::copy(paiArgs, std::back_inserter(argsVec));
+    llvm::copy(fasArgs, std::back_inserter(argsVec));
+
+    SILBuilderWithScope builder(fasi);
+    ApplyInst *appInst = builder.createApply(
+        /*Loc=*/fasi->getDebugLocation().getLocation(), /*Fn=*/callee,
+        /*Subs=*/pai->getSubstitutionMap(),
+        /*Args*/ argsVec,
+        /*isNonThrowing=*/fasi->isNonThrowing(),
+        /*SpecializationInfo=*/pai->getSpecializationInfo());
+    fasi->replaceAllUsesWith(appInst);
+    fasi->eraseFromParent();
+    erasedFullApply = true;
+  }
+  bool deletedDeadClosure = tryDeleteDeadClosure(pai);
+  return {erasedFullApply, deletedDeadClosure};
+}
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Entrypoint
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct MandatoryCombiner : SILFunctionTransform {
+  void run() override {
+    bool madeChange = false;
+    auto *f = getFunction();
+    for (auto &block : *f) {
+      for (auto ii = block.begin(), ie = block.end(); ii != ie;) {
+        auto *inst = &*ii;
+        // If any action is taken, the current instruction will be deleted.
+        // That instruction would be the definition of a partial apply.  Because
+        // the definition dominates all its uses, the previous instruction will
+        // be unaffected by the removal of the instruction and its uses.  So
+        // move the iterator back a step.  If no action is taken, it will be
+        // advanced twice.  If an action is taken, it will be advanced once to a
+        // new instruction (since the current one will have been deleted).
+        ii = prev_or_default(ii, block.begin(), block.end());
+
+        bool deleted = false;
+        if (auto *pai = dyn_cast<PartialApplyInst>(inst)) {
+          auto result = processPartialApply(pai);
+          deleted = result.second;
+          madeChange |= result.first || result.second;
+        }
+
+        if (deleted) {
+          // The deletion succeeded.  The iterator has already been moved back
+          // a step so as to avoid invalidation.  Advancing it one step will
+          // not advance it to the current instruction since that has been
+          // deleted.  Instead, it will advance it to the next *remaining*
+          // instruction after the previous instruction as desired.
+          if (ii != block.end()) {
+            ii = next_or_end(ii, block.end());
```

Comment:
I think the if else here (lines 152-156) can be simplified by using next_or_default. Specifically,

```
ii = next_or_default(ii, block.end() /*comparison*/, block.begin()/*default*/);
```

Reply:
Fixed.